### PR TITLE
s3ext: add gzip decompressing support

### PR DESCRIPTION
**Pull request by adam8157**
Now s3ext could recognize then decompress gzip encoded files
automatically, doesn't require any extra parameter, configuration or
extended filename.

### CODE REVIEW

File: gpAux/extensions/gps3ext/src/s3wrapper_test.cpp [hunk 1]
```diff
@@ -277,6 +290,13 @@ TEST(ExtWrapper, normal2_3segs) {
...
         64 * 1024, "00675684b6d6697571f22baaf407c6df", 2, 3, 64 * 1024 * 1024);
 }
+
+TEST(ExtWrapper, gzipped_normal2) {
+    ExtWrapperTest(
+        "https://s3-us-west-2.amazonaws.com/s3test.pivotal.io/dataset2/"
+        "gzipped_normal2/",
+        64 * 1024, "a930794bc885bccf6eed45bd40367a7d", 0, 1, 64 * 1024 * 1024);
+}
 #endif
```
Comment (lij55) [inline gpAux/extensions/gps3ext/src/s3wrapper_test.cpp]: What's the tested biggest zipped file size?

Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3wrapper_test.cpp]: after zipped, single file is ~200MB, bucket is ~1GB.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 2]
```diff
@@ -212,7 +223,13 @@ void *DownloadThreadfunc(void *data) {
...
 Downloader::Downloader(uint8_t part_num)
-    : num(part_num), o(NULL), chunkcount(0), readlen(0) {
+    : num(part_num),
+      o(NULL),
+      chunkcount(0),
+      readlen(0),
+      magic_bytes_num(0),
+      compression(S3_ZIP_NONE),
+      z_info(NULL) {
     this->threads = (pthread_t *)malloc(num * sizeof(pthread_t));
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: avoid to use malloc here

Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Any error handling for return NULL for malloc here?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 3]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
         pthread_create(&this->threads[i], NULL, DownloadThreadfunc,
                        this->buffers[i]);
     }
+
     readlen = 0;
     chunkcount = 0;
+    memset(this->magic_bytes, 0, 4);
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Use sizeof instead of 4 here.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 4]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
...
+
+    return true;
+}
+
+bool Downloader::set_compression() {
+    if ((this->magic_bytes[0] == 0x1f) && (this->magic_bytes[1] == 0x8b)) {
+        this->compression = S3_ZIP_GZIP;
+
+        this->z_info = new zstream_info();
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Any error handling for NULL?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 5]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
...
+        this->z_info->done_out = 0;
+        this->z_info->have_out = 0;
+    } else {
+        this->compression = S3_ZIP_NONE;
+    }
+
     return true;
 }
 
 bool Downloader::get(char *data, uint64_t &len) {
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Don't need to use uint64_t& but uint64_t

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 6]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
...
+    if (this->magic_bytes_num == 0) {
+        // get first 4(at least 2) bytes to check if this file is compressed
+        BlockingBuffer *buf = buffers[this->chunkcount % this->num];
+
+        if ((this->magic_bytes_num = buf->Read((char *)this->magic_bytes, 4)) >
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Use sizeof instead of 4

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 7]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
...
+    switch (this->compression) {
+        case S3_ZIP_GZIP:
+            return this->zstream_get(data, len);
+            break;
+        default:
+            return this->plain_get(data, len);
+    }
+}
+
+bool Downloader::plain_get(char *data, uint64_t &len) {
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Use uint64_t here

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 8]
```diff
@@ -250,23 +267,82 @@ bool Downloader::init(string url, string region, uint64_t size,
...
+    uint64_t tmplen = 0;
 
 RETRY:
+    // confirm there is no more available data, done with this file
     if (this->readlen == filelen) {
         len = 0;
         return true;
     }
 
     BlockingBuffer *buf = buffers[this->chunkcount % this->num];
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: I didn't see any check for assignment of num, which is assign by part_num. what happens if part_num = 0?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 9]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
     return true;
 }
 
+bool Downloader::zstream_get(char *data, uint64_t &len) {
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: As above

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 10]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
...
+    uint64_t filelen = this->o->Size();
+
+#define S3_ZIP_CHUNKSIZE 256 * 1024
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Specify 256 * 1024 here for any purpose?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 11]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
...
+            inflateEnd(strm);
+        }
+        len = 0;
+        return true;
+    }
+
+    BlockingBuffer *buf = buffers[this->chunkcount % this->num];
+
+    // strm is the structure used by zlib to decompress stream
+    if (!zinfo->inited) {
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: How about we use some initialize function here?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 12]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
...
         if (this->threads && this->threads[i])
             pthread_join(this->threads[i], NULL);
         if (this->buffers && this->buffers[i]) delete this->buffers[i];
     }
+
     if (this->o) delete this->o;
 }
 
 Downloader::~Downloader() {
     if (this->threads) free(this->threads);
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Do we need to call destroy here to if user don't call destroy directly?

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 13]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
...
     if (this->buffers) free(this->buffers);
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Mixed use of free and delete

File: gpAux/extensions/gps3ext/src/s3downloader.cpp [hunk 14]
```diff
@@ -286,21 +362,164 @@ bool Downloader::get(char *data, uint64_t &len) {
...
+                    inflateEnd(strm);
+                    return false;
+            }
+
+            zinfo->have_out = S3_ZIP_CHUNKSIZE - strm->avail_out;
+        }
+
+        // get another compressed chunk
+        // from magic_bytes, or buf->Read(), or both
+        if (!zinfo->have_out) {
```
Comment (macroyuyang) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Long function and indentions here

File: gpAux/extensions/gps3ext/src/s3downloader.cpp (hunk 2 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Yep, there is.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp (hunk 5 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Oh, that's what we meant to do, len will be the number of how much data it gets.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp (hunk 10 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: It's recommended by zlib, will add some comments here.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp (hunk 8 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: It's handled while reading from config file, min and max value are set to 1 & 8 right now.

File: gpAux/extensions/gps3ext/src/s3downloader.cpp (hunk 14 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/s3downloader.cpp]: Good point, planing to refactor get() in future because it's too much for this PR.

**Pull request merged**